feat(linter): inherit `rules` via the extended config files

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/9307

This gives the `extends` keyword some functionality: specifically to allow inheriting rules from other configuration files. Rules that are in the extended configuration will be used as the base, and then files in the config doing the extending will override the extended rules.

So, you can, for example, specify a base configuration that defines rules that you generally want to be enabled. Then, in each nested configuration file, you could disable rules depending on the directory:

```text
.oxlintrc.json       // { "rules: { "no-useless-escape": "error" } }
package1/
  .oxlintrc.json     // { "extends": ["../.oxlintrc.json"], "rules": { "no-useless-escape": "off" } }
package2/
  src                // uses ../.oxlintrc.json with `no-useless-escape` enabled 
```

As a side effect of this, building a config from a `.oxlintrc.json` file can result in config errors, even if all of the syntax and rules are configured correctly, because an extended configuration file might be incorrect or unparseable. For now, we simply raise an error if this occurs and stop execution, but in the future we could make this more graceful if we wanted and just ignore that file.

We will also need to revisit performance here, as I think a global cache config might be necessary so that we never load a file more than once and parse it more than once.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9308?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9308** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9308?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9308?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9217** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9217?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9156** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9156?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9155** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9155?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9153** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9152** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9308).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_linter/src/config/config_builder.rs:
```diff
@@ -675,4 +711,170 @@ mod test {
+    #[test]
+    fn test_extends_rules_single() {
+        let base_config =
+            config_store_from_path("../../apps/oxlint/fixtures/extends_config/rules_config.json");
```
NIT: I do not like that we are referencing files from another crate.

camchenry replied:
yeah I don't like that either. I've opened a PR on the top of the stack to fix this later: https://github.com/oxc-project/oxc/pull/9513

camchenry commented:
### Merge activity

* **Mar 3, 9:36 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
feat(linter): inherit `rules` via the extended config files (#9308)

- part of https://github.com/oxc-project/oxc/issues/9307

This gives the `extends` keyword some functionality: specifically to allow inheriting rules from other configuration files. Rules that are in the extended configuration will be used as the base, and then files in the config doing the extending will override the extended rules.

So, you can, for example, specify a base configuration that defines rules that you generally want to be enabled. Then, in each nested configuration file, you could disable rules depending on the directory:

```text
.oxlintrc.json       // { "rules: { "no-useless-escape": "error" } }
package1/
  .oxlintrc.json     // { "extends": ["../.oxlintrc.json"], "rules": { "no-useless-escape": "off" } }
package2/
  src                // uses ../.oxlintrc.json with `no-useless-escape` enabled
```

As a side effect of this, building a config from a `.oxlintrc.json` file can result in config errors, even if all of the syntax and rules are configured correctly, because an extended configuration file might be incorrect or unparseable. For now, we simply raise an error if this occurs and stop execution, but in the future we could make this more graceful if we wanted and just ignore that file.

We will also need to revisit performance here, as I think a global cache config might be necessary so that we never load a file more than once and parse it more than once.
